HV-1506 Regroup the Validator scoped properties in ValidationContext

marko-bekhta opened the pull request:
- https://hibernate.atlassian.net/browse/HV-1506

gsmet commented on engine/src/main/java/org/hibernate/validator/internal/engine/ValidationContext.java:
```diff
@@ -91,7 +90,7 @@
 
 	/**
-	 * The set of already processed unit of works. See {@link ProcessedUnit}.
+	 * The set of already processed unit of works. See {@code ProcessedUnit}.
```
Didn't you change that in another commit by pointing to the real classes?

gsmet commented on engine/src/main/java/org/hibernate/validator/internal/engine/ValidationContext.java:
```diff
@@ -121,19 +115,9 @@
-
-	/**
-	 * Clock provider which should be used in this context.
+	 * Context containing all validator level configuraions .
```
Typo in `configuraions`

gsmet commented on engine/src/main/java/org/hibernate/validator/internal/engine/ValidationContext.java:
```diff
@@ -459,44 +431,34 @@ private void markCurrentBeanAsProcessedForCurrentGroup(Object bean, Class<?> gro
-				ClockProvider clockProvider,
 				HibernateConstraintValidatorInitializationContext constraintValidatorInitializationContext,
-				boolean failFast) {
+				ValidatorContextProperties validatorContextProperties) {
```
Let's put it before the initialization context as it's more general.

gsmet commented on engine/src/main/java/org/hibernate/validator/internal/engine/ValidationContext.java:
```diff
@@ -459,44 +431,34 @@ private void markCurrentBeanAsProcessedForCurrentGroup(Object bean, Class<?> gro
 			this.constraintValidatorFactory = constraintValidatorFactory;
-			this.traversableResolver = traversableResolver;
-			this.clockProvider = clockProvider;
+			this.traversableResolver = validatorContextProperties.getWrapedWithCachingForSingleValidation();
```
I don't think it's a good idea to incorporate the `TraversableResolver` in it. It might not be `Validator` scoped so I don't think it's such a good idea to make it look like it.

gsmet commented on engine/src/main/java/org/hibernate/validator/internal/engine/ValidatorContextProperties.java:
```diff
@@ -0,0 +1,173 @@
+		return TraversableResolvers.wrapWithCachingForSingleValidation( traversableResolver, traversableResolverResultCacheEnabled );
+	}
+
+	public static class Builder {
```
Wondering if this builder pattern is bringing something to the plate. Looks like it's just noise. Moreover it might be error prone as we might forget to set up one of the necessary elements.

marko-bekhta replied:
well, I just don't like constructors with a huge amount of parameters :), but other than that, I think  there's not much benefit from it. Let me know if I should just remove it.

gsmet replied:
I tend to agree but I think it would be an error in this case. We instantiate the context in several places and with the builder, we could forget one of the arguments.

I think in this case, a multi-parameters constructor is best.

gsmet commented on engine/src/main/java/org/hibernate/validator/internal/engine/ValidatorContextProperties.java:
```diff
@@ -0,0 +1,173 @@
+ *
+ * @author Marko Bekhta
+ */
+public class ValidatorContextProperties {
```
I don't like the name but I can see how it is difficult to find a good name given we already have a `ValidatorContext` in `javax.validation`. I can't think of anything better than `ValidatorContext` though.

That being said, if we take into account my remark about limiting the class to `ValidatorImpl` and `ValidationContext`, I think that might be acceptable to name it `ValidatorContext` as the class will be very localized.

@gunnarmorling thoughts?

marko-bekhta replied:
Just needed to use something different than `ValidatorContext` :). How about `InstanceScopeValidatorContext` ?

gsmet replied:
Yuk :).

Maybe `ValidatorScopedContext`? And maybe we make it a package protected inner class of `ValidationContext`and reference it as `ValidationContext.ValidatorScopedContext`?

That's the best I can think of right now.

gsmet commented on engine/src/main/java/org/hibernate/validator/internal/engine/ValidatorContextProperties.java:
```diff
@@ -0,0 +1,173 @@
+	/**
+	 * The default traversable resolver for this factory.
+	 */
+	private final TraversableResolver traversableResolver;
```
As mentioned earlier, I think it would be a mistake to include this one. Let's keep it out of it to make it clear it can be validation scoped.

gsmet commented on engine/src/main/java/org/hibernate/validator/internal/engine/ValidatorContextProperties.java:
```diff
@@ -0,0 +1,173 @@
+		return this.traversableResolverResultCacheEnabled;
+	}
+
+	public TraversableResolver getWrapedWithCachingForSingleValidation() {
```
There are 2 p at `Wraped` but I think this method will go away.

gsmet commented on engine/src/main/java/org/hibernate/validator/internal/engine/ValidatorFactoryImpl.java:
```diff
@@ -69,42 +69,16 @@
  * @author Guillaume Smet
+ * @author Marko Bekhta
  */
 public class ValidatorFactoryImpl implements HibernateValidatorFactory {
```
I think I would have kept the changes to `ValidatorImpl` and I wouldn't have included `ValidatorFactoryImpl` and `ValidatorContextImpl` in the change. I would have built the context in the `ValidatorImpl` constructor.

marko-bekhta commented on engine/src/main/java/org/hibernate/validator/internal/engine/ValidationContext.java:
```diff
@@ -238,15 +210,15 @@ public ConstraintValidatorManager getConstraintValidatorManager() {
 	 */
 	public List<String> getParameterNames() {
-		if ( parameterNameProvider == null ) {
+		if ( executableParameters == null ) {
```
well ... honestly it bugs me, I mean not this particular change but Validation context. In particular two things:
- this null check was previously for the name provider as we were not passing it if the context wasn't for the parameters
- and `createConstraintViolation` method, this part in particular https://github.com/marko-bekhta/hibernate-validator/blob/d976b480573871c37c844ba8f44ffa17b268d9b6/engine/src/main/java/org/hibernate/validator/internal/engine/ValidationContext.java#L294 (If we have a `@NotNull` on return value and the value is actually null it will slip this if and fall to the last one. It is not visible in a constraint but still it kind of bugs me...)

So as I'm not sure about it I wanted to ask you - if it would be a good idea to split this context into 3 different ones ?

gsmet replied:
Yes, you're right, the test in `createConstraintViolation()` is buggy.

About separating `ValidationContext` in three contexts, I think it's too big of a change for a minor. We could maybe consider it for 6.1 but I'm not sure it's worth it.

I think what we are missing here is a `ValidationType` enum to define the type of validation currently at stake. Then we could base the conditions on the type and that would be totally safe. It would be a private member of `ValidationContext` so the change would be very local.

WDYT?

marko-bekhta replied:
sounds good. I'll add an enum for now and create a ticket for splitting the context, so we can decide later if it is worth it or not ...

gsmet replied:
Maybe, it's better to talk about "operation" instead of "type".

marko-bekhta commented:
I've rebased to current master and reorganized the commits a bit:
- First one contains only changes for `ValidatorScopedContext` with the fixes to the things you've mentioned.
- In the next one I've cleaned up a constructor of `ValidationFactory` so now all configuration property extraction should follow the same pattern.
- And in the last one - I've added `ValidatioOperation` enum as were suggested.

gunnarmorling commented on engine/src/main/java/org/hibernate/validator/internal/engine/ValidatorContextImpl.java:
```diff
@@ -188,4 +188,87 @@ public Validator getValidator() {
+	 * Context object storing all possible validator configurable properties.
+	 * There should be just one per {@link Validator} instance.
+	 */
+	static class ValidatorScopedContext {
```
Maybe just `ValidatorContext`? And how about adding all the things scoped to the VF, too (or having another, dedicated context for that)? That way we'd get back to much leaner signatures of the methods passing through these objects.

gsmet replied:
We already have `ValidatorContext` in the spec. That's why we tried to have another name.

@marko-bekhta my idea was to put this one in `ValidationContext` not `ValidatorContextImpl`. I don't think it has anything to do with `ValidatorContextImpl`.

gsmet replied:
> And how about adding all the things scoped to the VF, too (or having another, dedicated context for that)? That way we'd get back to much leaner signatures of the methods passing through these objects.

I haven't thought about this yet. But, anyway, let's keep it for another PR.

marko-bekhta replied:
@gsmet that was my mistake - classes are together and I dropped the new one into a wrong one. 
@gunnarmorling +1 for updating the factory as well. Let's do it in another PR.

gunnarmorling commented on engine/src/main/java/org/hibernate/validator/internal/engine/ValidationContext.java:
```diff
@@ -238,15 +220,15 @@ public ConstraintValidatorManager getConstraintValidatorManager() {
 	 */
 	public List<String> getParameterNames() {
-		if ( parameterNameProvider == null ) {
+		if ( !ValidationOperation.PARAMETER_VALIDATION.equals( validationOperation ) ) {
```
Or even raise an exception? It seems calling this method in this case would indicate a programming error?

gsmet replied:
+1

marko-bekhta replied:
This is, kind of, related to suggestion to split the `ValidationContext` into multiple ones (https://hibernate.atlassian.net/browse/HV-1526). The way how this method is used:
```java
// create a constraint validator context
ConstraintValidatorContextImpl constraintValidatorContext = new ConstraintValidatorContextImpl(
		validationContext.getParameterNames(),
		validationContext.getClockProvider(),
		valueContext.getPropertyPath(),
		descriptor
);
```
wouldn't allow to just throw an exception here. So I was thinking that if we create different kinds of `ValidationContext`s we could then also have different implementations of `HibernateConstraintValidatorContext` and then we wouldn't need such kind of checks here, as well as in `ValidationContext#createConstraintViolation()`.

gsmet commented:
Added a commit with cosmetic improvements and comment fixes and merged. Thanks!
